Remove coming soon settings option for v2

lsl opened the pull request:
#### Changes proposed in this Pull Request

* Splits out the settings page changes from #45606 
* Reworks settings page to only consider v1 values and the v2 transition

#### Testing instructions

Regardless of etk's/0-sandbox.php define: `define( 'WPCOM_PUBLIC_COMING_SOON', true );`

* `http://calypso.localhost:3000/settings/general/<yourtestsite>.wo\rdpress.com?flags=coming-soon-v2`

Should show existing coming soon settings, switching to public/private and saving settings should remove the option.

* `http://calypso.localhost:3000/settings/general/<yourtestsite>.wo\rdpress.com?flags=coming-soon-v2=false`

Should leave things functioning the same as:

* `http://wordpress.com/settings/general/<yourtestsite>.wo\rdpress.com`

props: @roo2 / @ramonjd (alot of code got squashed/rebased out of #45606

a8ci18n commented on client/my-sites/site-settings/form-general.jsx:
```diff
@@ -158,7 +158,7 @@ export class SiteSettingsFormGeneral extends Component {
 				<Button href={ '/domains/add/' + siteSlug } onClick={ this.trackUpgradeClick }>
 					<Gridicon icon="plus" />{ ' ' }
-					{ translate( 'Add custom address', { context: 'Site address, domain' } ) }
+					{ translate( 'Add custom address', { comment: 'Site address, domain' } ) }
```
:warning: This change will be queued for retranslation. We'll display the English text until then.
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **2** times:
`translate( 'Add a custom address' )` *ES Score: 7*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/remove-coming-soon)

ramonjd commented on client/lib/signup/step-actions/index.js:
```diff
@@ -191,7 +192,11 @@ function getNewSiteParams( {
 	};
 
-	newSiteParams.options.wpcom_coming_soon = getNewSiteComingSoonSetting( state );
+	if ( config.isEnabled( 'coming-soon-v2' ) ) {
```
We're using `'editing-toolkit/coming-soon'` over in https://github.com/Automattic/wp-calypso/pull/45892/files#diff-08c61c7c946b2c05d7687b49ef35985bR108

Pulled from the original https://github.com/Automattic/wp-calypso/pull/45606/files#diff-4663f7e9182cfd61c9e36163548cfdddR369

I don't have a strong opinion on the flag name. Just noting it so we can update in the appropriate places after we settle on which term we use.

ramonjd replied:
I've updated my branch to use `coming-soon-v2` in https://github.com/Automattic/wp-calypso/pull/45892/commits/e927805008cc3a55c64e05c13150044839e567eb

ramonjd commented on client/my-sites/site-settings/form-general.jsx:
```diff
@@ -377,38 +402,46 @@ export class SiteSettingsFormGeneral extends Component {
+
+	privacyRadioComingSoon() {
+		const { fields, isRequestingSettings, eventTracker, translate } = this.props;
+		const wasComingSoon = parseInt( fields.was_wpcom_coming_soon, 10 ) === 1;
```
Does `was_wpcom_coming_soon` mean that a site was formerly v1?

lsl replied:
Yup, this only supports v1. v2 settings are entirely ignored, the only thing enabling v2 does is switches on this functionality to disappear the UI option after a v1 user switches out of the setting. `was_wpcom_coming_soon` is just to store the intermediate state so it doesn't disappear until the user clicks save. Without `was_...` the UI element would disappear as soon as you change the option (but don't yet save.)

ramonjd commented on client/my-sites/site-settings/form-general.jsx:
```diff
@@ -377,38 +402,46 @@ export class SiteSettingsFormGeneral extends Component {
+		const wasComingSoon = parseInt( fields.was_wpcom_coming_soon, 10 ) === 1;
+		const isComingSoon = parseInt( fields.wpcom_coming_soon, 10 ) === 1;
+
+		// If v2 and not currently coming soon, hide
```
Just to elaborate on the comment and get it straight in my mind, if:

1. coming soon v2 is activated and 
2. the user's site is **not** running v1 mode 

we remove the ability to toggle coming soon altogether. This includes new sites.

lsl replied:
I think - Yes. v2 sites are just public sites with a behind the scenes site option.

lsl replied:
Also - iirc during prelaunch these settings are disabled.

lsl commented:
Rebased, reverted the settings page and pushed up a bit of a commit cleanup to make changes more clear: https://github.com/Automattic/wp-calypso/pull/45896/files/4ff37845dac8bb5ea137069842ad2a98274e8ce9..44aa622177e8a3cb693ca5368cc685083405e24a.

Split the settings page out in #46034 minus all the refactor cleanup, that should be safe to merge early as it doesn't change anything unless the flag is flipped on.

There is a fair bit still in here to do with launching / site creation steps. 

Should we be looking at moving that stuff serverside into headstart by chance?

ramonjd commented:
> There is a fair bit still in here to do with launching / site creation steps. Should we be looking at moving that stuff serverside into headstart by chance?

@roo2 has something working over at https://github.com/Automattic/wp-calypso/pull/45950 using hooks, effectively bypassing headstart 🕺 

As for launching, there's also an action we could hook into when a user launches a site, in the `launch_site` lib

lsl committed:
wip

p-jackson commented on client/blocks/eligibility-warnings/index.tsx:
```diff
@@ -249,6 +249,7 @@ const mapDispatchToProps = {
 		saveSiteSettings( selectedSiteId, {
 			blog_public: 1,
 			wpcom_coming_soon: 0,
+			wpcom_public_coming_soon: 0,
```
I don't think we should change the setting when the site goes atomic. We did this for v1 because coming soon would break things on atomic. But the idea with v2 is that sites aren't private to begin with, so don't need to be taken public.

Unless you're thinking this is just temporary until we've worked out syncing of options with atomic?

lsl replied:
Sounds good, will remove / add a comment on this.

lsl replied:
As discussed leaving this for now as I think it makes sense given the function name we need to figure out why / where this is being called if it is no longer needed.

p-jackson replied:
Tracked this down I think. The only time `makeSitePublic()` is called is if the list of AT holds includes `SITE_NOT_PUBLIC`. The list of holds usually comes from `/sites/:id/automated-transfers/eligibility`, however `SITE_NOT_PUBLIC` doesn't exist on the backend. I couldn't find it in the history either (although I didn't search exhaustively).

So under normal circumstances (like attempting to install a plugin or generate ftp credentials) the `<EligibilityWarnings>` UI doesn't call `makeSitePublic()` and set `wpcom_coming_soon` to `0`.

I did find something in Calypso that's reusing the `<EligibilityWarnings>` UI though, and it passes in a hard-coded `SITE_NOT_PUBLIC` prop in order to get the right copy and button behaviour.
If your site is _already atomic_ and you go to `/settings/performance/:siteSlug` in Calypso you'll see this message, which is just a `<EligibilityWarnings>` under the hood.
<img width="737" alt="Screenshot 2020-10-05 at 11 06 10 AM" src="https://user-images.githubusercontent.com/1500769/95028336-0926e000-06fc-11eb-831b-348a611ebfc1.png">

I believe this is because you can only set up Jetpack performance features if your site is public. Privacy prevents it from being able to cache your pages and images I suppose.

So I don't think we should be setting `wpcom_public_coming_soon` to `0` in this case. In this context `makeSitePublic()` really just means `makeSiteWorkWithJetpackPerformance()`, and it should work for any value of `wpcom_public_coming_soon`.

lsl replied:
Thanks this makes sense, I'll add a comment here in its place as I don't think this will be clear in code without it.

p-jackson replied:
> I don't think this will be clear in code without it
lol yup

p-jackson commented on client/state/selectors/is-site-coming-soon.js:
```diff
@@ -27,8 +22,11 @@ export default function isSiteComingSoon( state, siteId ) {
-		// Site settings returns a numerical value for wpcom_coming_soon.
-		return parseInt( settings.wpcom_coming_soon, 10 ) === 1;
+		// Site settings returns a numerical value for wpcom_public_coming_soon.
+		return (
+			parseInt( settings.wpcom_coming_soon, 10 ) === 1 ||
+			parseInt( settings.wpcom_public_coming_soon, 10 ) === 1
+		);
```
This is a tough choice. Having the `isSiteComingSoon` selector return true for either v1 or v2 makes sense because it means it'll often just show the sort of copy in the UI that we want (the feature is still named "Coming Soon" after all).

But `isSiteComingSoon` is also used for showing notices about why SEO and podcasting settings don't work. And those are only applicable for v1. Ultimately those SEO and podcasting messages never show because of other booleans `if`ing them out. But using `isSiteComingSoon` for both v1 and v2 sort of makes the logic even more confusing than it currently is.

On the other hand, it'd be sort of gross to have a `_v1` and `_v2` version of the same selector.

Sorry, I don't really know which way I'm advocating tbh.

lsl replied:
I'll take a look at where this is used, I'm more than ok with a v2 suffix but those particular examples look like work arounds for v1 shortcomings rather than coming soon feature specific things.

In that case it is probably better to v2 flag check on those work arounds individually.

lsl replied:
Duplicated this function into a V2 version and audited the usages of V1, added v2 where it made sense. Could use some eyes on those changes - they're all untested and I'm not really sure how to get to some of the states I've touched in this PR now.

p-jackson commented on client/my-sites/site-settings/form-general.jsx:
```diff
@@ -485,15 +490,12 @@ export class SiteSettingsFormGeneral extends Component {
 
 		return (
-			<>
+			<Fragment>
```
Did I read somewhere your editor was having trouble with the `<>` shorthand? Is that what's going on here?

lsl committed:
Clean up public / v2 options on new sites

lsl committed:
Rollback some test changes, add v2 ones

lsl committed:
V2 coming soon check

lsl committed:
Comingsoon v2 warning for seo settings

lsl committed:
Missed the not on rollback

p-jackson commented on client/state/selectors/get-new-site-coming-soon-setting.ts:
```diff
@@ -12,3 +12,7 @@ import shouldNewSiteBePrivateByDefault from './should-new-site-be-private-by-def
 }
+
+export function getNewSiteComingSoonSettingV2( state: object ): number {
+	return shouldNewSiteBePrivateByDefault( state ) ? 0 : 1;
```
I'm reluctant to use `shouldNewSiteBePrivateByDefault` at all inside this function. That's because coming-soon-v2 has nothing to do with private-by-default, and the future of private-by-default affect v2 in any way. In fact v2 is an attempt to try to disentangle coming-soon from privacy. 

I know it doesn't make any difference to the actual outcome (since `shouldNewSiteBePrivateByDefault()` is hard-coded to return `true`) but I think it will further tie the two things together in peoples minds.

p-jackson replied:
I changed the logic here so that I was able to create a coming-soon-v2 site using the `/start` flow. The above logic was basically hard-coded to always return 0.

lsl replied:
This was a bit of hand wringing by me. I was trying to keep the code sort of similar to v1 but I think it made things way more complicated, I've taken your change and gone further by dropping the function completely and just using the feature flag directly. I think its still a bit confusing but if we do a sweep with your visibility enum we can probably drop the one line functions that just return constant values completely anyway.

p-jackson commented on client/my-sites/site-settings/seo-settings/form.jsx:
```diff
@@ -312,30 +313,37 @@ export class SeoForm extends React.Component {
-						text={ ( function () {
-							if ( isSitePrivate ) {
-								if ( siteIsComingSoon ) {
+				{ ( isSitePrivate || isSiteHidden || siteIsComingSoonV2 ) &&
```
Haven't tested yet, but I was thinking that maybe we don't need `siteIsComingSoonV2` here, since it's already covered by `isSiteHidden`.

If the user has a coming-soon-v2 site, and they've somehow manauged to change their `blog_public` setting to `1`, then there's no reason to prevent them from changing SEO settings. Google should happily index their coming-soon-v2 site.

lsl replied:
While coming soon v2 currently means site is hidden I didn't want to make the assumption since they're different options and have different text below.

p-jackson committed:
Make it possible to create coming-soon-v2 sites using /start

The only thing that determines whether a site should be created as
coming-soon-v2 is whether the feature flag is enabled. Not whether the
site is private or not.

This change also makes it possible for `getNewSitecomingSoonSettingV2()`
to return 1. Before it would always return 0 no matter what.

p-jackson committed:
Fix tests and simplify diff in unit test file

This PR doesn't change the behaviour of `isSiteComingSoon` and I think
this commit should have returned the tests back to testing the original
behaviour.

I've also tried to reduce the PR diff to make it a bit easier to read by
putting some of the original code back and creating a seperate
`describe` block for `isSiteComingSoonV2`.

p-jackson commented:
There's good stuff in this PR, but I think we should write a clearer mission statement for what it is and isn't trying to achieve. That way it's going to be easier to know when this is mergeable. Can I propose a purpose of:

1. Make it possible to create coming-soon-v2 sites using the `/start` flow
2. Update the general settings form so that coming-soon-v1 sites can't be put back into coming soon mode after launch (only when the feature flag is enabled, we won't ship to customers yet)
3. Hide/disable any Calypso pages or settings that don't work when a site is in coming soon v2 mode (if any)

w.r.t (3) I'm thinking that most things will actually just work for v2 sites and nothing needs to be hidden, but investigating this is in scope for this PR

w.r.t (2) this is the unclearest part of the PR for me. Have I even correctly summarised how we want the general settings form to work?

lsl commented:
> There's good stuff in this PR, but I think we should write a clearer mission statement for what it is and isn't trying to achieve. That way it's going to be easier to know when this is mergeable.

Most of this is just iterations on the original PR ramon drafted #45167 with a few things pulled out so far, keen to split out more mergable parts if you can spot what they are.

> w.r.t (2) this is the unclearest part of the PR for me. Have I even correctly summarised how we want the general settings form to work? 

Currently,

If the v2 **flag** is enabled the coming soon  settings page option will disappear once a user switches out of coming soon and hits save. It will show for any user with the coming soon v1 site option. v2 site **option** is completely ignored by the settings page.

With pbxlJb-q6-p2#comment-538,

This is changing in #46119, users will be able to change back into coming soon mode regardless of version, once v2 is enabled switching back and forth will upgrade your site to v2/public mode.

p-jackson commented:
> Currently, If the v2 flag is enabled the coming soon settings page option will disappear once a user switches out of coming soon and hits save.

> users will be able to change back into coming soon mode regardless of version, once v2 is enabled switching back and forth will upgrade your site to v2/public mode.

Cool! Thanks for the clarification. I think this is a good plan. I was initially unclear whether coming-soon-v2 would re-use the same radio controls as v1, and this makes it clear that we currently don't expect the UI to change much between v1 and v2.

lsl committed:
makeSitePublic change to comment

lsl committed:
Remove getNewSiteComingSoonSettingV2 - was overcomplicating where it was being used

lsl committed:
Missed the other usage
